### PR TITLE
Fix private key immutability error when using OpenSSL 3.0 

### PR DESCRIPTION
**Pull request by debreczeni**
fixes "OpenSSL::PKey::PKeyError: pkeys are immutable on OpenSSL 3.0" Error.
Details: https://github.com/GemHQ/money-tree/issues/63
Although no separate test case have been written, all tests pass OK.
Related to heroku-22 stack.

### CODE REVIEW

**Comment by bdescamps** [comment]
I have some error on the openssl_extensions_spec.rb file.
Maybe you should use ```::OpenSSL::PKey::EC.generate("secp256k1")``` instead of ```OpenSSL::PKey::EC.new("secp256k1").generate_key```

**Comment by debreczeni** [comment]
> Hello @debreczeni, I have some error on the openssl_extensions_spec.rb file. Maybe you should use `::OpenSSL::PKey::EC.generate("secp256k1")` instead of `OpenSSL::PKey::EC.new("secp256k1").generate_key`

Thank you for the catch. Which version of OpenSSL are you running? It should have an OpenSSL version check as this may only be available in v3. The specs run OK with OpenSSL v3.0.5. I will have to double-check on that.

**Comment by bdescamps** [comment]
I am on OpenSSL 3.0.2. It is strange that it works, because  when using ```new``̀` combine with ```generate_key``` to me you break the immutable aspect of openssl 3.0.

**Comment by q9f** [comment]
This works with OpenSSL 1.1, so I'm happy to accept this. Can you run Rufo?

**Comment by bdescamps** [comment]
I think we should use the runner ubuntu-22.04 instead on ubuntu-latest (which is not latest but 20.04)  in github spec file : 
```[.github/workflows/spec.yml ]```

**Comment by bdescamps** [comment]
I submit pull request on your repo : https://github.com/debreczeni/money-tree/pulls . Fix Rufo /  Adding pipeline on ubuntu 22.04 / FIx rspec test on generate

**Comment by q9f** [comment]
> I think we should use the runner ubuntu-22.04 instead on ubuntu-latest (which is not latest but 20.04) in github spec file : `[.github/workflows/spec.yml ]`

TIL https://github.com/actions/runner-images/issues/6399

**Comment by bdescamps** [comment]
Maybe we can flag with 22.04 until latest <> 22.04, and rollback after

**Comment by debreczeni** [comment]
> I submit pull request on your repo : https://github.com/debreczeni/money-tree/pulls . Fix Rufo / Adding pipeline on ubuntu 22.04 / FIx rspec test on generate

Thank you. I have rebased the PR branch to GemHQ/master then added your commits from your pull request (to avoid criss-cross merge commits in git. If you rebase on my original commit as well I can do a merge commit from your PR if you'd like).

**Comment by q9f** [comment]
Done, it fails on Ubuntu 22.04 with Ruby 3.1.

**Comment by q9f** [comment]
Can you rebase on master? ref #68 

Interestingly it only affects Ubuntu 22.4 and Ruby 3.1

**Comment by bdescamps** [comment]
It's relative to this issue : https://github.com/GemHQ/money-tree/issues/62 . Everything is linked :hot_face:

**Comment by q9f** [comment]
@debreczeni can you confirm this works locally for you?

**Comment by bdescamps** [comment]
I have just submitted, work to activate the legacy sect on ubuntu 22-04.
@debreczeni can you check the PR on your repositories ?

**Comment by bdescamps** [comment]
Any update about this feature ? @debreczeni @q9f .  If it's more simply / quick we can merge this one, and i will create another PR to fix spec. As you want  !

**Comment by debreczeni** [comment]
> @debreczeni can you confirm this works locally for you?

Yes it works as in specs run OK within the gem and it works in the project used as a gem as well, successfully initializing Key classes from existing private keys (which was the original reason why this fix came up) I have merged/included all commits from @bdescamps  then now rebased to master again and it works there as well with OpenSSL 3.0.5 on Mac OS.

**Comment by debreczeni** [comment]
> I have just submitted, work to activate the legacy sect on ubuntu 22-04. @debreczeni can you check the PR on your repositories ? (https://github.com/debreczeni/money-tree/pull/2/files)

Thank you, merged then rebased to master, everything still seems OK on my end.

**Comment by bdescamps** [comment]
> > I have just submitted, work to activate the legacy sect on ubuntu 22-04. @debreczeni can you check the PR on your repositories ? (https://github.com/debreczeni/money-tree/pull/2/files)
> 
> Thank you, merged then rebased to master, everything still seems OK on my end.

Good news : Hope we will have soon a new money-tree release :)

File: .github/workflows/spec.yml [hunk 1]
```diff
@@ -23,6 +23,12 @@ jobs:
       with:
         ruby-version: ${{ matrix.ruby }}
         bundler-cache: false
+    - name: Load OpenSSL ripemd160
+      run: |
+        sed -i '/^\default = default_sect/a legacy = legacy_sect' /etc/ssl/openssl.cnf
+        sed -i '/^\[default_sect\]/a activate = 1' /etc/ssl/openssl.cnf
+        echo "[legacy_sect]" >> /etc/ssl/openssl.cnf
+        echo "activate = 1" >> /etc/ssl/openssl.cnf
```
Comment (q9f) [inline .github/workflows/spec.yml:31]:
```suggestion
        sudo echo "activate = 1" >> /etc/ssl/openssl.cnf
```

File: .github/workflows/spec.yml [hunk 2]
```diff
@@ -23,6 +23,12 @@ jobs:
       with:
         ruby-version: ${{ matrix.ruby }}
         bundler-cache: false
+    - name: Load OpenSSL ripemd160
+      run: |
+        sed -i '/^\default = default_sect/a legacy = legacy_sect' /etc/ssl/openssl.cnf
```
Comment (q9f) [inline .github/workflows/spec.yml:28]:
```suggestion
        sudo sed -i '/^\default = default_sect/a legacy = legacy_sect' /etc/ssl/openssl.cnf
```

File: .github/workflows/spec.yml [hunk 3]
```diff
@@ -23,6 +23,12 @@ jobs:
...
+        sed -i '/^\[default_sect\]/a activate = 1' /etc/ssl/openssl.cnf
```
Comment (q9f) [inline .github/workflows/spec.yml:29]:
```suggestion
        sudo sed -i '/^\[default_sect\]/a activate = 1' /etc/ssl/openssl.cnf
```

File: .github/workflows/spec.yml [hunk 4]
```diff
@@ -23,6 +23,12 @@ jobs:
...
+        echo "[legacy_sect]" >> /etc/ssl/openssl.cnf
```
Comment (q9f) [inline .github/workflows/spec.yml:30]:
```suggestion
        sudo echo "[legacy_sect]" >> /etc/ssl/openssl.cnf
```

**Review by q9f — COMMENTED** [review]
LGTM but it fails for lacking permissions. You can just use sudo without password in GA.

**Comment by q9f** [comment]
😩

**Comment by bdescamps** [comment]
> weary

I'm using act locally to simulate github action and have no pb ( sed and echo commands are execute as root) 
Maybe we can try using ```sudo su``` before

``` 
        whoami
        ls -lah 
        sudo su
        sed -i '/^\default = default_sect/a legacy = legacy_sect' /etc/ssl/openssl.cnf
        sed -i '/^\[default_sect\]/a activate = 1' /etc/ssl/openssl.cnf
        echo "[legacy_sect]" >> /etc/ssl/openssl.cnf
        echo "activate = 1" >> /etc/ssl/openssl.cnf
 ```

**Comment by q9f** [comment]
There is really no hope for Ubuntu. Anyone using Ubuntu should pressure the maintainers to upgrade openssl to 3.0.7 or later or stay on openssl 1.1. Or, alternatively, use the custom provided `config/openssl.conf`.

Everyone else (Archlinux, MacOS, etc.) should be fine. @debreczeni I included your patch in #71 - thanks!

**Pull request closed without merging**